test/unit-test/ota_os_posix_utest.c

```diff
@@ -32,48 +32,36 @@
 #include <string.h>
 #include <mqueue.h>
 #include <time.h>
+#include <unistd.h>
 #include "unity.h"
 
 /* For accessing OTA private functions and error codes. */
+#include "ota.h"
 #include "ota_os_posix.h"
-#include "ota_private.h"
-
 
 /* Testing constants. */
-#define MAX_MESSAGES           10 /*!< Maximum number of messages in the event queue. */
 #define TIMER_NAME             "dummy_name"
-#define OTA_DEFAULT_TIMEOUT    10 /*!< Timeout in seconds. */
-
-/* Timer used in os_posix.c */
-extern timer_t otaTimer;
+#define OTA_DEFAULT_TIMEOUT    1000 /*!< Timeout in milliseconds. */
 
 /* Interfaces for Timer and Event. */
+static OtaTimerId_t timer_id = 0;
 static OtaTimerInterface_t timer;
 static OtaEventInterface_t event;
-static OtaTimerContext_t * pTimerContext = NULL;
 static OtaEventContext_t * pEventContext = NULL;
+static struct timespec start;
+static bool timerCallbackInovked = false;
 
-/**
- * @brief Get the Time elapsed from the timer.
- *
- * This is used to ensure that the timer has started successfully,
- * by using the timer id otaTimer to get the time elapsed and
- * store it into timer structure.
- *
- * @return long time elapsed in nano seconds.
- */
-static long getTimeElapsed()
+static void timerCallback()
 {
-    struct itimerspec timerAttr;
-    long retVal = 0;
+    struct timespec now;
 
-    /* On error, -1 is returned else 0. */
-    if( timer_gettime( otaTimer, &timerAttr ) == 0 )
-    {
-        retVal = timerAttr.it_value.tv_nsec;
-    }
+    timerCallbackInovked = true;
+    clock_gettime( CLOCK_REALTIME, &now );
+    int elapse_ms = ( now.tv_sec - start.tv_sec ) * 1e3 + ( now.tv_nsec - start.tv_nsec ) / 1e6;
 
-    return retVal;
+    /* Assert the time elapsed is within 0.9 to 1.1 of the specified time. */
+    TEST_ASSERT_LESS_THAN( OTA_DEFAULT_TIMEOUT * 1.1, elapse_ms );
+    TEST_ASSERT_GREATER_THAN( OTA_DEFAULT_TIMEOUT * 0.9, elapse_ms );
```

Comment:
What is this testing?  It looks brittle.

Reply:
It's testing the timer implementation, one of the test starts the timer with timeout of 1000 ms, this is asserting the time elapsed is between 900 ms and 1100 ms. Before  the test requires direct access to the static variable defined in os posix, so I changed to this way. If you have other suggestion I would like to hear.

Reply:
On the face of it, it looks reasonable, and yet this test will run on systems that are probably not real-time, and may indeed be virtual.  Also, this provides only a single sample of the latency of a single timer.  If timer latency is a real concern, then this test is not thorough enough.  If it's not a real concern, then why test it?

Reply:
This test only runs on linux, it's for posix implementation only.

Reply:
I agree, but that doesn't answer my question.